### PR TITLE
[SPARK-22787] [TEST] [SQL] Add a TPC-H query suite

### PR DESCRIPTION
**Pull request by gatorsmile**
## What changes were proposed in this pull request?
Add a test suite to ensure all the TPC-H queries can be successfully analyzed, optimized and compiled without hitting the max iteration threshold.

## How was this patch tested?
N/A

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/text/TextFileFormat.scala [hunk 1]
```diff
@@ -116,9 +116,10 @@ class TextFileFormat extends TextBasedFileFormat with DataSourceRegister {
     readToUnsafeMem(broadcastedHadoopConf, requiredSchema, textOptions.wholeText)
   }
 
-  private def readToUnsafeMem(conf: Broadcast[SerializableConfiguration],
-      requiredSchema: StructType, wholeTextMode: Boolean):
-  (PartitionedFile) => Iterator[UnsafeRow] = {
+  private def readToUnsafeMem(
+      conf: Broadcast[SerializableConfiguration],
+      requiredSchema: StructType,
+      wholeTextMode: Boolean): (PartitionedFile) => Iterator[UnsafeRow] = {
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/text/TextFileFormat.scala]: This is just a style fix of a PR I merged today.

**Comment by gatorsmile** [comment]
cc @cloud-fan @maropu @kiszk

File: sql/core/src/test/resources/tpch/q1.sql [hunk 2]
```diff
@@ -0,0 +1,23 @@
+-- using default substitutions
+
+select
+	l_returnflag,
```
Comment (dongjoon-hyun) [inline sql/core/src/test/resources/tpch/q1.sql]: Can we use space instead of `tab`?

Comment (gatorsmile) [reply sql/core/src/test/resources/tpch/q1.sql]: This is the generated SQL. I plan to keep them unchanged.

File: sql/core/src/test/scala/org/apache/spark/sql/TPCHQuerySuite.scala [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        |`l_shipinstruct` STRING, `l_shipmode` STRING, `l_comment` STRING)
+        |USING parquet
+      """.stripMargin)
+  }
+
+  val tpchQueries = Seq(
+    "q1", "q2", "q3", "q4", "q5", "q6", "q7", "q8", "q9", "q10", "q11",
+    "q12", "q13", "q14", "q15", "q16", "q17", "q18", "q19", "q20", "q21", "q22")
+
+  private def checkGeneratedCode(plan: SparkPlan): Unit = {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/TPCHQuerySuite.scala]: How about making a base trait to remove duplicate code between this and `TPCDSQuerySuite`?

**Comment by maropu** [comment]
LGTM for one minor comment.
BTW, how about also adding star schema benchmark (many join cases) to check compilation?

**Comment by gatorsmile** [comment]
https://www.cs.umb.edu/~poneil/StarSchemaB.PDF This one?

**Comment by maropu** [comment]
yea.

**Comment by maropu** [comment]
Probably, we could get sql queries from here https://github.com/electrum/ssb-dbgen (I don't look into yet though...)

**Comment by kiszk** [comment]
LGTM

File: sql/core/src/test/scala/org/apache/spark/sql/TPCHQuerySuite.scala [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+/**
+ * This test suite ensures all the TPC-H queries can be successfully analyzed, optimized
+ * and compiled without hitting the max iteration threshold.
+ */
+class TPCHQuerySuite extends BenchmarkQueryTest {
+
+  /**
+   * Drop all the tables
+   */
+  protected override def afterAll(): Unit = {
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/TPCHQuerySuite.scala]: As same as `TPCDSQuerySuite`, we can use `BenchmarkQueryTest.afterAll`.

**Comment by viirya** [comment]
LGTM with one minor comment.

File: sql/core/src/test/scala/org/apache/spark/sql/BenchmarkQueryTest.scala [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  // the max iteration of analyzer/optimizer batches.
+  assert(Utils.isTesting, "spark.testing is not set to true")
+
+  /**
+   * Drop all the tables
+   */
+  protected override def afterAll(): Unit = {
+    try {
+      // For debugging dump some statistics about how much time was spent in various optimizer rules
+      logWarning(RuleExecutor.dumpTimeSpent())
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/BenchmarkQueryTest.scala]: Why warning for debug uses?

Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/BenchmarkQueryTest.scala]: Also, is that a bad idea to dump the time for each query?

File: sql/core/src/test/scala/org/apache/spark/sql/TPCHQuerySuite.scala (hunk 4 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/TPCHQuerySuite.scala]: yea we don't need to overwrite it here.

File: sql/core/src/test/scala/org/apache/spark/sql/BenchmarkQueryTest.scala (hunk 5 above)
Comment (gatorsmile) [reply sql/core/src/test/scala/org/apache/spark/sql/BenchmarkQueryTest.scala]: If we do use logWarning, the messages will not be shown in the test log.

Comment (gatorsmile) [reply sql/core/src/test/scala/org/apache/spark/sql/BenchmarkQueryTest.scala]: This is just to give the overall picture how long each rule takes. 

I plan to submit another PR to track which rule takes an effect for a specific query and also record the time cost.

Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/BenchmarkQueryTest.scala]: oh, ok

**Comment by gatorsmile** [comment]
cc @maropu Feel free to submit a PR for adding SSB

**Comment by maropu** [comment]
yea, sure. I have much bandwidth now :)

**Comment by gatorsmile** [comment]
Thanks! Merged to master.

**Comment by maropu** [comment]
https://github.com/apache/spark/pull/19990

**Comment by maropu** [comment]
@gatorsmile Any progress on this? https://github.com/apache/spark/pull/19982#discussion_r157119941
After I thought your comment, I came up with collecting metrics for each rule like;
https://github.com/apache/spark/compare/master...maropu:MetricSpike
This conflicts with your activity, or this is not acceptable? welcome any comment.

**Comment by gatorsmile** [comment]
@maropu Thanks for your contribution. It looks over engineering. We do not need such complicated solutions for this simple use case. We just need to record them in the log. We are also proposing new APIs for our logs. @jiangxb1987 is working on the design.

**Comment by maropu** [comment]
ok, I just look forward to the proposal.

**Pull request closed without merging**